### PR TITLE
Concat, minify and auto inject javascript sources into index.html

### PR DESCRIPTION
**Pull request by pcenta**
Workflow for concatenating and minifying javascript source files in correct order into one file and auto injecting minified/non-minified sources depending on build environment into index.html.

### CODE REVIEW

**Comment by team-pct** [comment]
Can one of the admins verify this patch?

**Comment by team-pct** [comment]
Can you explain what is the benefits ?

**Comment by pcenta** [comment]
@team-pct 
This takes all javascript source files, joins them in one file and then minifies it. Result is that only one file gets loaded as source javascript without whitespaces, comments and unused variables. It increases performance (not a gamechanger).
Negative would be that it complicates understanding what's going on from user screenshot since source will always be one file and variables will be renamed.

**Comment by rafaell-lycan** [comment]
That's really good. IMO the only thing I'd change is the 3rd party been hardcoded on the `index.html`. We could take advantage of cache to avoid inject it every single update, and also isolate it on a distinct build to be injected as well. With that, it will be simple and transparent the way to not only update the dependencies and cleaner regarding loading only 2 assets, plus, it can be minified as well.

File: package.json [hunk 1]
```diff
@@ -105,9 +105,13 @@
     "gulp": "^4.0.2",
     "gulp-filter": "^5.1.0",
```
Comment (rafaell-lycan) [inline package.json:106]: The only break change on this dep was the drop support for the Node versions below 8. WDYT?

File: package.json [hunk 2]
```diff
@@ -105,9 +105,13 @@
...
     "gulp-gzip": "^1.4.2",
+    "gulp-inject": "^5.0.5",
     "gulp-jsbeautifier": "^2.1.2",
     "gulp-jshint": "^2.0.0",
     "gulp-load-plugins": "^1.5.0",
```
Comment (rafaell-lycan) [inline package.json:111]: Same deal here, as this version has some vulnerability and they fix it on 1.6.0, but still, the only major difference here set a minimum Node version. WDYT?

File: package.json (hunk 1 above)
Comment (pcenta) [reply package.json:106]: Not sure what you're referring to. Final changelog has no changes in marked lines.

File: package.json (hunk 2 above)
Comment (pcenta) [reply package.json:111]: Same as above, final changes to package.json are only about adding needed dependencies. Existing versions are unchanged.

Comment (rafaell-lycan) [reply package.json:111]: No problem, it's your call, just to let you know once you're changing a bit gulp dependencies as well.

https://github.com/jackfranklin/gulp-load-plugins#160

File: package.json (hunk 1 above)
Comment (rafaell-lycan) [reply package.json:106]: No problem, it's your call.

File: gulpfile.js [hunk 3]
```diff
@@ -150,6 +235,48 @@ const nw = new nwBuilder({
 /*************
  * gulp tasks *
  *************/
+gulp.task('clean-js', (done) => {
+  del(['src/app/js/*.js'], {force: true}).then(done());
```
Comment (ArcticZeroo) [inline gulpfile.js:239]: `.then(done)` or `.then(() => done());`. this calls done immediately

Comment (pcenta) [reply gulpfile.js:239]: Indeed, my bad for missing that.

**Comment by rafaell-lycan** [comment]
@pcenta could you make those last changes and resolve the conflicts on this PR?

**Comment by pcenta** [comment]
> @pcenta could you make those last changes and resolve the conflicts on this PR?

I'll take a look at it. Should be quite straight forward.

**Comment by pcenta** [comment]
Moved to https://github.com/popcorn-official/popcorn-desktop/pull/1691.

**Pull request closed without merging**